clang/labs/lab1/kashirin_alexander/Rename_Plugin.cpp

```diff
@@ -0,0 +1,196 @@
+        rewriter.ReplaceText(constructor->getLocation(), cur_name.size(),
+                             new_name);
+      }
+    }
+    return true;
+  }
+
+  bool VisitCXXNewExpr(clang::CXXNewExpr *newExpr) {
+    if (type == IdType::Class) {
+      if (newExpr->getConstructExpr()->getType().getAsString() == cur_name) {
+        rewriter.ReplaceText(newExpr->getExprLoc(), cur_name.size() + 4,
+                             "new " + new_name);
+      }
+    }
+    return true;
+  }
+
+  bool save_changes() { return rewriter.overwriteChangedFiles(); }
+
+private:
+  clang::Rewriter rewriter;
+  IdType type;
+  std::string cur_name;
+  std::string new_name;
+};
+
+class RenameASTConsumer : public clang::ASTConsumer {
+public:
+  explicit RenameASTConsumer(clang::CompilerInstance &CI, IdType type,
+                             clang::StringRef cur_name,
+                             clang::StringRef new_name)
+      : Visitor(clang::Rewriter(CI.getSourceManager(), CI.getLangOpts()), type,
+                cur_name, new_name) {}
+
+  void HandleTranslationUnit(clang::ASTContext &context) override {
+    Visitor.TraverseDecl(context.getTranslationUnitDecl());
+    if (Visitor.save_changes()) {
+      llvm::errs() << "An error occurred while saving changes to a file!\n";
+    }
+  }
+
+private:
+  RenameVisitor Visitor;
+};
+
+class RenamePlugin : public clang::PluginASTAction {
+public:
+  std::unique_ptr<clang::ASTConsumer>
+  CreateASTConsumer(clang::CompilerInstance &CI,
+                    clang::StringRef InFile) override {
+    return std::make_unique<RenameASTConsumer>(CI, type, cur_name, new_name);
+  }
+
+protected:
+  bool ParseArgs(const clang::CompilerInstance &CI,
+                 const std::vector<std::string> &args) override {
+    std::vector<std::pair<std::string, std::string>> params = {
+        {"type=", ""}, {"cur-name=", ""}, {"new-name=", ""}};
+
+    if (!args.empty() && args[0] == "help") {
+      PrintHelp(llvm::errs());
+      return true;
+    }
+
+    if (args.size() < 3) {
+      PrintParamsError(CI);
+      return false;
+    }
+
+    for (const auto &arg : args) {
+      bool is_found = false;
+      for (auto &param : params) {
+        if (arg.find(param.first) == 0 && param.second.empty()) {
+          param.second = arg.substr(param.first.size());
+          is_found = true;
+          break;
+        }
+      }
+      if (!is_found) {
+        PrintParamsError(CI);
+        return false;
+      }
+    }
+
+    std::vector<std::pair<std::string, IdType>> id_type = {
+        {"var", IdType::Var}, {"func", IdType::Func}, {"class", IdType::Class}};
+    size_t i;
+    for (i = 0; i < id_type.size(); i++) {
+      if (params[0].second == id_type[i].first) {
+        type = id_type[i].second;
+        break;
+      }
+    }
+    if (i == id_type.size()) {
+      PrintParamsError(CI);
+      return false;
+    }
+    cur_name = params[1].second;
+    new_name = params[2].second;
```

Comment:
Add the check for names to comply with standard and filter out unacceptable names, e.g.
- foo - valid
- foo.123 - not valid
- 1foo - not valid
- foo__bAR - valid
- ^foo - not valid
and so on

Reply:
checks added

Reply:
Please, point out where exactly

Reply:
Rename_Plugin.cpp: 
 bool isValidIdentifier(const std::string &name) const {
    clang::Lexer lexer(clang::SourceLocation(), clang::LangOptions(),
                       name.c_str(), name.c_str(),
                       name.c_str() + name.length());
    for (clang::Token tok; !lexer.LexFromRawLexer(tok);)
      if (tok.isNot(clang::tok::identifier))
        return false;
    return true;
  }
this function is to check for names for compliance with standard